### PR TITLE
Add and fix a baseline set of warnings

### PR DESCRIPTION
**Pull request by lopopolo**
- Add a reasonable default set of warnings to `lib.rs`.
- Warn on missing unsafe documentation via clippy.
- Warn on unsafe op in unsafe fn.
- Add `RUSTFLAGS="-D warnings"` to `check` job in CI
- Add rust2018 idioms lint and fix warnings.
- Add `clippy::cargo` lint group and add missing package metadata to the haiku platform crate.

I've only tested these changes on macOS, and due to the `path` pragma we use, I can't test these changes on other platforms.

### CODE REVIEW

**Review by astraw — APPROVED** [review]
Thanks, looks good to me!

**Comment by lopopolo** [comment]
Holding off on this until https://github.com/strawlab/iana-time-zone/pull/75 is merged, then I will resolve conflicts.

**Comment by astraw** [comment]
Thanks, this is great. From my side you are welcome to merge when ready.

**Comment by lopopolo** [comment]
let me rebase to pull in the recently merged changes and then I'll merge once CI runs.

**Comment by lopopolo** [comment]
Only macOS MSRV and android nightly cross builds have failed, which are known issues. Merging.

**Pull request merged**
